src/main/java/io/cdap/plugin/cdc/DMLFlattener.java

```diff
@@ -0,0 +1,124 @@
+import io.cdap.cdap.api.plugin.PluginConfig;
+import io.cdap.cdap.etl.api.Emitter;
+import io.cdap.cdap.etl.api.PipelineConfigurer;
+import io.cdap.cdap.etl.api.Transform;
+import io.cdap.cdap.etl.api.TransformContext;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+/**
+ * Extracts the DML record from the output of a cdc source for direct manipulation.
+ */
+@Plugin(type = Transform.PLUGIN_TYPE)
+@Name("DMLFlattener")
+@Description("Flattens DML records output by a CDC source.")
+public class DMLFlattener extends Transform<StructuredRecord, StructuredRecord> {
+  private static final String OP_TYPE = "CDC_OP_TYPE";
+  private static final String CHANGE_TRACKING_VERSION = "CHANGE_TRACKING_VERSION";
+  private final Conf conf;
+  private Map<Schema, Schema> schemaCache;
+  private Schema configuredOutputSchema;
+  private boolean addOpType = false;
+  private boolean addTrackingVersion = false;
+
+  public DMLFlattener(Conf conf) {
+    this.conf = conf;
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    if (conf.schema != null) {
+      try {
+        pipelineConfigurer.getStageConfigurer().setOutputSchema(Schema.parseJson(conf.schema));
+      } catch (IOException e) {
+        throw new IllegalArgumentException("Unable to parse configured schema: " + e.getMessage(), e);
+      }
+    }
+  }
+
+  @Override
+  public void initialize(TransformContext context) throws IOException {
+    configuredOutputSchema = conf.schema == null ? null : Schema.parseJson(conf.schema);
+    addOpType = configuredOutputSchema.getField(OP_TYPE) != null;
+    addTrackingVersion = configuredOutputSchema.getField(CHANGE_TRACKING_VERSION) != null;
+    schemaCache = new HashMap<>();
+  }
+
+  @Override
+  public void transform(StructuredRecord record, Emitter<StructuredRecord> emitter) throws Exception {
+    StructuredRecord dml = record.get("dml");
+    if (dml == null) {
+      return;
+    }
+
+    Schema rowSchema = Schema.parseJson((String) dml.get("rows_schema"));
+    Schema outputSchema = schemaCache.computeIfAbsent(rowSchema, this::createOutputSchema);
+
+    StructuredRecord.Builder output = StructuredRecord.builder(outputSchema);
+    if (addOpType) {
+      output.set(OP_TYPE, dml.get("op_type").toString());
+    }
+    if (addTrackingVersion) {
+      output.set(CHANGE_TRACKING_VERSION, dml.get("change_tracking_version"));
+    }
+    Map<String, Object> valueMap = dml.get("rows_values");
+    if (valueMap == null) {
+      valueMap = new HashMap<>();
+    }
+    for (Map.Entry<String, Object> entry : valueMap.entrySet()) {
+      output.set(entry.getKey(), entry.getValue());
+    }
+    emitter.emit(output.build());
+  }
+
+  private Schema createOutputSchema(Schema rowSchema) {
+    List<Schema.Field> fields = new ArrayList<>(rowSchema.getFields().size() + 2);
```

Comment:
nit: maybe add comment for op_type and change_tracking_version to know why +2 is needed